examples/read_pem_files/main.rs

```diff
@@ -0,0 +1,69 @@
+use config::{Config, Format, Map, Value, ValueKind, FileStoredFormat};
+use std::io::{Error, ErrorKind};
+
+/// The private and public key sources will be read into their associated variable:
+#[derive(serde::Deserialize, Clone, Debug)]
+pub struct Settings {
+    pub private_key: Option<String>,
+    pub public_key: Option<String>,
+}
+
+fn main() {
+    // Sourcing from two separate files for the `Settings` struct,
+    // these are both marked as optional to avoid failure if the file doesn't exist.
+    let file_public_key = config::File::new("public.pem", PemFile).required(false);
+    let file_private_key = config::File::new("private.pem", PemFile).required(false);
+
+    // Provide the sources and build the config object:
+    let settings = Config::builder()
+        .add_source(public_key)
+        .add_source(private_key)
+        .build()
+        .unwrap();
+
+    // Deserialize the config object into your Settings struct:
+    let settings: Settings = settings.try_deserialize().unwrap();
+    println!("{:#?}", settings);
+}
+
+#[derive(Debug, Clone)]
+pub struct PemFile;
+
+impl Format for PemFile {
+    fn parse(
+        &self,
+        uri: Option<&String>,
+        text: &str,
+    ) -> Result<Map<String, config::Value>, Box<dyn std::error::Error + Send + Sync>> {
+        // Store any valid keys into this map, they'll be merged with other sources into the final config map:
+        let mut result = Map::new();
+
+        // Identify the PEM encoded data type by the first occurrence found:
+        // NOTE: This example is kept simple, multiple or other encoded types are not handled.
+        let key_type = ["PUBLIC", "PRIVATE"].into_iter().find(|s| text.contains(s));
```

Comment:
Your code seams be correct. Unfortunately, I have not been able to compile it without some errors... I cannot understand why...  It only works if I change to this:
```suggestion
        let key_type = Vec::<&str>::from(["PUBLIC", "PRIVATE"]).into_iter().find(|s| text.contains(s));
```

It gives this error: ```expected a `Fn<(char,)>` closure, found `str`
the trait `Fn<(char,)>` is not implemented for `str`
the following other types implement trait `Pattern<'a>`:
  &'b str
  &'c &'b str```

and on
```Some("PRIVATE") => "private_key",```
```mismatched types
expected reference `&&str`
   found reference `&'static str```

Reply:
Due to the way I ran and modified your example, this turned out to be due to the `edition` in `Cargo.toml`. I used `2021` while `config-rs` is still set to `2018`. It turns out that making an array into an iter is one of the improvements that 2021 edition provides.

We can avoid the typing and from syntax here by using the `vec!` macro for now:

```suggestion
        let key_type = vec!["PUBLIC", "PRIVATE"].into_iter().find(|s| text.contains(s));
```